File: jdroid-java-core/src/test/java/com/jdroid/java/marshaller/MarshallerProviderTest.java
package com.jdroid.java.marshaller;

import com.jdroid.java.collections.Lists;
import com.jdroid.java.collections.Maps;
import com.jdroid.java.json.JsonMap;

import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class MarshallerProviderTest {
	
	private MarshallerProvider marshallerProvider;
	
	public MarshallerProviderTest() {
		marshallerProvider = MarshallerProvider.get();
		marshallerProvider.addMarshaller(DummyClass.class, new DummyClassMarshaller());
	}
	
	@DataProvider
	public Iterator<Object[]> marshallDataProvider() {
		List<Object[]> cases = Lists.newArrayList();
		cases.add(new Object[] { 1, "1" });
		cases.add(new Object[] { "1", "1" });
		cases.add(new Object[] { Lists.newArrayList(1, 2, 3), "[1,2,3]" });
		cases.add(new Object[] { Lists.newArrayList("1", "2", "3"), "[\"1\",\"2\",\"3\"]" });
		
		DummyClass dummyClass = new DummyClass();
		dummyClass.stringProperty = "1";
		String dummyJson = "{\"stringProperty\":\"1\"}";
		cases.add(new Object[] { dummyClass, dummyJson });
		cases.add(new Object[] { Lists.newArrayList(dummyClass, dummyClass), "[" + dummyJson + "," + dummyJson + "]" });

		dummyClass = new DummyClass();
		dummyClass.longProperty = 2L;
		dummyJson = "{\"longProperty\":2}";
		cases.add(new Object[] { dummyClass, dummyJson });
		cases.add(new Object[] { Lists.newArrayList(dummyClass, dummyClass), "[" + dummyJson + "," + dummyJson + "]" });

		dummyClass = new DummyClass();
		dummyClass.stringListProperty = Lists.newArrayList("3", "4");
		dummyJson = "{\"stringListProperty\":[\"3\",\"4\"]}";
		cases.add(new Object[] { Lists.newArrayList(dummyClass, dummyClass), "[" + dummyJson + "," + dummyJson + "]" });

		dummyClass = new DummyClass();
		dummyClass.longListProperty = Lists.newArrayList(5L, 6L);
		dummyJson = "{\"longListProperty\":[5,6]}";
		cases.add(new Object[] { dummyClass, dummyJson });
		cases.add(new Object[] { Lists.newArrayList(dummyClass, dummyClass), "[" + dummyJson + "," + dummyJson + "]" });

		dummyClass = new DummyClass();
		dummyClass.stringMapProperty = Maps.newLinkedHashMap();
		dummyJson = "{}";
		cases.add(new Object[] { dummyClass, dummyJson });
		cases.add(new Object[] { Lists.newArrayList(dummyClass, dummyClass), "[" + dummyJson + "," + dummyJson + "]" });

		dummyClass = new DummyClass();
		dummyClass.stringMapProperty = Maps.newLinkedHashMap();
		dummyClass.stringMapProperty.put("a", "b");
		dummyJson = "{\"stringMapProperty\":{\"a\":\"b\"}}";
		cases.add(new Object[] { dummyClass, dummyJson });
		cases.add(new Object[] { Lists.newArrayList(dummyClass, dummyClass), "[" + dummyJson + "," + dummyJson + "]" });

		dummyClass = new DummyClass();
		dummyClass.longMapProperty = Maps.newLinkedHashMap();
		dummyClass.longMapProperty.put("a", 1L);
		dummyJson = "{\"longMapProperty\":{\"a\":1}}";
		cases.add(new Object[] { dummyClass, dummyJson });
		cases.add(new Object[] { Lists.newArrayList(dummyClass, dummyClass), "[" + dummyJson + "," + dummyJson + "]" });

		dummyClass = new DummyClass();
		dummyClass.dummyClassMapProperty = Maps.newLinkedHashMap();

		DummyClass innerDummyClass = new DummyClass();
		innerDummyClass.stringProperty = "2";
		innerDummyClass.longProperty = 1L;

		dummyClass.dummyClassMapProperty.put("a", innerDummyClass);
		dummyJson = "{\"dummyClassMapProperty\":{\"a\":{\"longProperty\":1,\"stringProperty\":\"2\"}}}";
		cases.add(new Object[] { dummyClass, dummyJson });
		cases.add(new Object[] { Lists.newArrayList(dummyClass, dummyClass), "[" + dummyJson + "," + dummyJson + "]" });

		cases.add(new Object[] { new DummyClass(), "{}" });
		
		return cases.iterator();
	}
	
	@Test(dataProvider = "marshallDataProvider")
	public void marshall(Object data, String expectedJson) {
		String result = marshallerProvider.marshall(data, null, null).toString();
		Assert.assertEquals(result.replace(" ", ""), expectedJson);
	}
	
	private class DummyClass {
		
		protected String stringProperty;
		protected Long longProperty;
		protected List<String> stringListProperty;
		protected List<Long> longListProperty;
		protected Map<String, String> stringMapProperty;
		protected Map<String, Long> longMapProperty;
		protected Map<String, DummyClass> dummyClassMapProperty;
	}
	
	public class DummyClassMarshaller implements Marshaller<DummyClass, JsonMap> {
		
		@Override
		public JsonMap marshall(DummyClass dummyClass, MarshallerMode mode, Map<String, String> extras) {
			JsonMap map = new JsonMap(mode, extras);
			map.put("stringProperty", dummyClass.stringProperty);
			map.put("longProperty", dummyClass.longProperty);
			map.put("stringListProperty", dummyClass.stringListProperty);
			map.put("longListProperty", dummyClass.longListProperty);
			map.put("stringMapProperty", dummyClass.stringMapProperty);
			map.put("longMapProperty", dummyClass.longMapProperty);
			map.put("dummyClassMapProperty", dummyClass.dummyClassMapProperty);
			return map;
		}
	}
}

File: jdroid-java-core/src/test/java/com/jdroid/java/marshaller/MarshallerProviderTest.kt
package com.jdroid.java.marshaller

import com.jdroid.java.collections.Lists
import com.jdroid.java.collections.Maps
import com.jdroid.java.json.JsonMap

import org.testng.Assert
import org.testng.annotations.DataProvider
import org.testng.annotations.Test

import java.util.HashMap

class MarshallerProviderTest {

    private val marshallerProvider: MarshallerProvider = MarshallerProvider.get()

    init {
        marshallerProvider.addMarshaller(DummyClass::class.java, DummyClassMarshaller())
    }

    @DataProvider
    fun marshallDataProvider(): Iterator<Array<Any>> {
        val cases = Lists.newArrayList<Array<Any>>()
        cases.add(arrayOf(1, "1"))
        cases.add(arrayOf("1", "1"))
        cases.add(arrayOf(Lists.newArrayList(1, 2, 3), "[1,2,3]"))
        cases.add(arrayOf(Lists.newArrayList("1", "2", "3"), "[\"1\",\"2\",\"3\"]"))

        var dummyClass = DummyClass()
        dummyClass.stringProperty = "1"
        var dummyJson = "{\"stringProperty\":\"1\"}"
        cases.add(arrayOf(dummyClass, dummyJson))
        cases.add(arrayOf(Lists.newArrayList(dummyClass, dummyClass), "[$dummyJson,$dummyJson]"))

        dummyClass = DummyClass()
        dummyClass.longProperty = 2L
        dummyJson = "{\"longProperty\":2}"
        cases.add(arrayOf(dummyClass, dummyJson))
        cases.add(arrayOf(Lists.newArrayList(dummyClass, dummyClass), "[$dummyJson,$dummyJson]"))

        dummyClass = DummyClass()
        dummyClass.stringListProperty = Lists.newArrayList("3", "4")
        dummyJson = "{\"stringListProperty\":[\"3\",\"4\"]}"
        cases.add(arrayOf(Lists.newArrayList(dummyClass, dummyClass), "[$dummyJson,$dummyJson]"))

        dummyClass = DummyClass()
        dummyClass.longListProperty = Lists.newArrayList(5L, 6L)
        dummyJson = "{\"longListProperty\":[5,6]}"
        cases.add(arrayOf(dummyClass, dummyJson))
        cases.add(arrayOf(Lists.newArrayList(dummyClass, dummyClass), "[$dummyJson,$dummyJson]"))

        dummyClass = DummyClass()
        dummyClass.stringMapProperty = Maps.newLinkedHashMap()
        dummyJson = "{}"
        cases.add(arrayOf(dummyClass, dummyJson))
        cases.add(arrayOf(Lists.newArrayList(dummyClass, dummyClass), "[$dummyJson,$dummyJson]"))

        dummyClass = DummyClass()
        dummyClass.stringMapProperty = Maps.newLinkedHashMap()
        dummyClass.stringMapProperty!!["a"] = "b"
        dummyJson = "{\"stringMapProperty\":{\"a\":\"b\"}}"
        cases.add(arrayOf(dummyClass, dummyJson))
        cases.add(arrayOf(Lists.newArrayList(dummyClass, dummyClass), "[$dummyJson,$dummyJson]"))

        val map = HashMap<String, String>()
        map["a"] = "b"
        map["c"] = "d"
        dummyJson = "{\"a\":\"b\",\"c\":\"d\"}"
        cases.add(arrayOf(map, dummyJson))
        cases.add(arrayOf(Lists.newArrayList<Map<String, String>>(map, map), "[$dummyJson,$dummyJson]"))

        dummyClass = DummyClass()
        dummyClass.longMapProperty = Maps.newLinkedHashMap()
        dummyClass.longMapProperty!!["a"] = 1L
        dummyJson = "{\"longMapProperty\":{\"a\":1}}"
        cases.add(arrayOf(dummyClass, dummyJson))
        cases.add(arrayOf(Lists.newArrayList(dummyClass, dummyClass), "[$dummyJson,$dummyJson]"))

        dummyClass = DummyClass()
        dummyClass.dummyClassMapProperty = Maps.newLinkedHashMap()

        val innerDummyClass = DummyClass()
        innerDummyClass.stringProperty = "2"
        innerDummyClass.longProperty = 1L

        dummyClass.dummyClassMapProperty!!["a"] = innerDummyClass
        dummyJson = "{\"dummyClassMapProperty\":{\"a\":{\"longProperty\":1,\"stringProperty\":\"2\"}}}"
        cases.add(arrayOf(dummyClass, dummyJson))
        cases.add(arrayOf(Lists.newArrayList(dummyClass, dummyClass), "[$dummyJson,$dummyJson]"))

        cases.add(arrayOf(DummyClass(), "{}"))

        return cases.iterator()
    }

    @Test(dataProvider = "marshallDataProvider")
    fun marshall(data: Any, expectedJson: String) {
        val result = marshallerProvider.marshall(data, null, null)!!.toString()
        Assert.assertEquals(result.replace(" ", ""), expectedJson)
    }

    inner class DummyClass {

        var stringProperty: String? = null
        var longProperty: Long? = null
        var stringListProperty: List<String>? = null
        var longListProperty: List<Long>? = null
        var stringMapProperty: MutableMap<String, String>? = null
        var longMapProperty: MutableMap<String, Long>? = null
        var dummyClassMapProperty: MutableMap<String, DummyClass>? = null
    }

    inner class DummyClassMarshaller : Marshaller<DummyClass, JsonMap> {

        override fun marshall(value: DummyClass?, mode: MarshallerMode?, extras: Map<String, String>?): JsonMap? {
            val map = JsonMap(mode, extras)
            map["stringProperty"] = value!!.stringProperty
            map["longProperty"] = value.longProperty
            map["stringListProperty"] = value.stringListProperty
            map["longListProperty"] = value.longListProperty
            map["stringMapProperty"] = value.stringMapProperty
            map["longMapProperty"] = value.longMapProperty
            map["dummyClassMapProperty"] = value.dummyClassMapProperty
            return map
        }
    }
}
